File: iOS/Session/Sources/Session/CurrentAccount/CurrentAccountCoordinator.swift
import Foundation

/// Application triggers that may start a current-account refresh.
public enum CurrentAccountRefreshTrigger: Sendable {
    /// First authenticated launch of the process; refreshes once regardless of staleness.
    case coldStart
    /// The app returned to the foreground; refreshes only when the snapshot is stale.
    case foreground
    /// Profile opened; refreshes only when the snapshot is stale.
    case profileOpened
}

/// How long a successful validation keeps the current account fresh.
public let currentAccountFreshnessInterval: TimeInterval = 5 * 60

/// Application-scoped owner of the signed-in account. Serializes sign-in installation,
/// server refresh, username mutation, and logout clearing so an older remote result can
/// never overwrite newer local state, and publishes one replaying observable value that
/// every account consumer observes instead of keeping manual copies.
///
/// State changes are applied synchronously inside the actor and fenced by an epoch:
/// any operation that began before a mutation or logout finds the epoch advanced and
/// discards its result. Disk writes are enqueued in state-change order so the persisted
/// snapshot can never run ahead of, or reorder against, the published value.
public actor CurrentAccountCoordinator {
    private let accountStorage: AccountStorage
    private let authenticationRepository: AuthenticationRepository
    private let now: @Sendable () -> Date
    private let freshnessInterval: TimeInterval

    private var snapshot: CurrentAccountSnapshot?
    private var epoch = 0
    private var hydrationTask: Task<Void, Never>?
    private var refreshTask: Task<Void, Never>?
    private var hasRefreshedAfterColdStart = false
    private var persistenceChain: Task<Void, Never>?
    private var continuations: [UUID: AsyncStream<AccountBundle?>.Continuation] = [:]

    init(
        accountStorage: AccountStorage,
        authenticationRepository: AuthenticationRepository,
        now: @escaping @Sendable () -> Date = { Date() },
        freshnessInterval: TimeInterval = currentAccountFreshnessInterval
    ) {
        self.accountStorage = accountStorage
        self.authenticationRepository = authenticationRepository
        self.now = now
        self.freshnessInterval = freshnessInterval
    }

    /// The current signed-in account, or nil when signed out or not yet hydrated.
    public var account: AccountBundle? { snapshot?.account }

    /// Replays the current account immediately, then yields every subsequent change.
    public func accountUpdates() -> AsyncStream<AccountBundle?> {
        AsyncStream { continuation in
            let id = UUID()
            continuations[id] = continuation
            continuation.yield(snapshot?.account)
            continuation.onTermination = { @Sendable _ in
                Task { [weak self] in await self?.removeContinuation(id) }
            }
        }
    }

    /// Loads the persisted snapshot once and publishes it before any remote work,
    /// so authenticated routing never waits on a network request.
    @discardableResult
    public func hydrate() async -> AccountBundle? {
        if hydrationTask == nil {
            hydrationTask = Task { await self.performHydration() }
        }
        await hydrationTask?.value
        return snapshot?.account
    }

    /// Installs the canonical account produced by sign-in. Throws when the snapshot could
    /// not be persisted, so sign-in can report the failure instead of silently producing a
    /// session that will not survive the next launch.
    public func install(account: AccountBundle) async throws {
        epoch += 1
        let newSnapshot = CurrentAccountSnapshot(account: account, validatedAt: now())
        snapshot = newSnapshot
        hasRefreshedAfterColdStart = true
        publish()
        try await persistOrdered { [accountStorage] in
            try await accountStorage.store(snapshot: newSnapshot)
        }
    }

    /// Refreshes from the server when the trigger is eligible. Concurrent triggers
    /// coalesce into one request; a failure keeps the snapshot published and stale so a
    /// later trigger retries.
    public func refresh(trigger: CurrentAccountRefreshTrigger) async {
        guard snapshot != nil else { return }
        switch trigger {
        case .coldStart:
            guard !hasRefreshedAfterColdStart else { return }
        case .foreground, .profileOpened:
            guard isStale else { return }
        }
        hasRefreshedAfterColdStart = true
        if refreshTask == nil {
            refreshTask = Task { await self.performRefresh() }
        }
        let task = refreshTask
        await task?.value
    }

    /// PATCHes the username and, on success, persists and publishes the submitted value.
    /// A refresh that began earlier can no longer overwrite it; a failure leaves the
    /// published account unchanged.
    public func updateUsername(_ username: String) async -> Result<String, Error> {
        guard let current = snapshot else {
            return .failure(UpdateUsernameError.noStoredAccount)
        }
        let startEpoch = epoch
        let result = await authenticationRepository.updateUsername(
            accountId: current.account.accountId,
            username: username
        )
        switch result {
        case .success:
            if epoch == startEpoch, let live = snapshot {
                epoch += 1
                let updated = CurrentAccountSnapshot(
                    account: live.account.withUsername(username),
                    validatedAt: live.validatedAt
                )
                snapshot = updated
                publish()
                enqueuePersistence { [accountStorage] in
                    try? await accountStorage.store(snapshot: updated)
                }
            }
            return .success(username)
        case .failure(let error):
            return .failure(error)
        }
    }

    /// Clears the account on logout; any in-flight refresh or mutation result can no
    /// longer restore it. Throws when the persisted snapshot could not be removed, so
    /// callers can surface an incomplete logout instead of leaving the account
    /// recoverable on disk.
    public func clear() async throws {
        epoch += 1
        snapshot = nil
        hasRefreshedAfterColdStart = false
        publish()
        try await persistOrdered { [accountStorage] in
            try await accountStorage.delete()
        }
    }

    /// Awaits every persistence operation enqueued so far.
    func awaitPersistence() async {
        await persistenceChain?.value
    }

    private var isStale: Bool {
        guard let validatedAt = snapshot?.validatedAt else { return true }
        return now().timeIntervalSince(validatedAt) >= freshnessInterval
    }

    private func performHydration() async {
        let startEpoch = epoch
        let persisted = (try? await accountStorage.retrieve()) ?? nil
        guard let persisted, epoch == startEpoch, snapshot == nil else { return }
        snapshot = persisted
        publish()
    }

    private func performRefresh() async {
        defer { refreshTask = nil }
        let startEpoch = epoch
        let result = await authenticationRepository.getCurrentAccount()
        guard epoch == startEpoch, snapshot != nil else { return }
        guard case .success(let account) = result else { return }
        let newSnapshot = CurrentAccountSnapshot(account: account, validatedAt: now())
        snapshot = newSnapshot
        publish()
        enqueuePersistence { [accountStorage] in
            try? await accountStorage.store(snapshot: newSnapshot)
        }
    }

    private func publish() {
        let account = snapshot?.account
        for continuation in continuations.values {
            continuation.yield(account)
        }
    }

    private func removeContinuation(_ id: UUID) {
        continuations[id] = nil
    }

    private func enqueuePersistence(_ operation: @escaping @Sendable () async -> Void) {
        let previous = persistenceChain
        persistenceChain = Task {
            await previous?.value
            await operation()
        }
    }

    /// Enqueues a persistence operation in state-change order and awaits it, rethrowing
    /// its failure. Refresh and username persists stay fire-and-forget (the server is
    /// authoritative and the next refresh self-heals); install and clear must be durable.
    private func persistOrdered(_ operation: @escaping @Sendable () async throws -> Void) async throws {
        let previous = persistenceChain
        let task = Task {
            await previous?.value
            try await operation()
        }
        persistenceChain = Task { _ = try? await task.value }
        try await task.value
    }
}

File: Android/session/src/main/java/com/felipearpa/tyche/session/CurrentAccountCoordinator.kt
@file:OptIn(ExperimentalTime::class)

package com.felipearpa.tyche.session

import com.felipearpa.tyche.session.authentication.domain.AuthenticationRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlin.time.Clock
import kotlin.time.Duration
import kotlin.time.Duration.Companion.minutes
import kotlin.time.ExperimentalTime

enum class CurrentAccountRefreshTrigger { COLD_START, FOREGROUND, PROFILE_OPENED }

/**
 * Single configuration point for how long a successful current-account validation stays fresh.
 */
val currentAccountFreshness: Duration = 5.minutes

/**
 * Application-scoped coordinator that owns the observable signed-in account.
 *
 * It serializes sign-in installation, refresh, username mutation, and logout clearing so an
 * older remote result can never overwrite a newer local change: every state change happens
 * under one mutex while network calls run outside it, and each in-flight operation is fenced
 * by the epoch it started under. Refresh follows stale-while-revalidate — the persisted
 * snapshot is published immediately, one coalesced request reconciles it when a trigger finds
 * it stale, and a failed refresh retains the snapshot for a later retry.
 */
class CurrentAccountCoordinator internal constructor(
    private val accountStorage: AccountStorage,
    private val authenticationRepository: AuthenticationRepository,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default),
    private val nowEpochMillis: () -> Long = { Clock.System.now().toEpochMilliseconds() },
    private val freshness: Duration = currentAccountFreshness,
) {
    private val mutex = Mutex()
    private var snapshot: CurrentAccountSnapshot? = null
    private var epoch = 0
    private var hasRefreshedAfterColdStart = false
    private var hydration: Deferred<Unit>? = null
    private var refreshJob: Deferred<Unit>? = null

    private val mutableState = MutableStateFlow<AccountBundle?>(null)
    val state: StateFlow<AccountBundle?> = mutableState.asStateFlow()

    /**
     * Publishes the persisted snapshot, if any, and returns the current account. Concurrent
     * calls coalesce into one storage read.
     */
    suspend fun hydrate(): AccountBundle? {
        val pending = mutex.withLock {
            hydration ?: scope.async { performHydration() }.also { hydration = it }
        }
        pending.await()
        return state.value
    }

    private suspend fun performHydration() {
        val startEpoch = mutex.withLock { epoch }
        val persisted = runCatching { accountStorage.retrieve() }.getOrNull()
        mutex.withLock {
            if (persisted != null && epoch == startEpoch && snapshot == null) {
                snapshot = persisted
                mutableState.value = persisted.account
            }
        }
    }

    /**
     * Installs a just-signed-in account as freshly validated; sign-in already reconciled it
     * with the server, so the next cold-start refresh is suppressed.
     *
     * A storage failure propagates to the sign-in use case — losing the snapshot silently
     * would sign the user in for one process lifetime only. The write itself cannot be
     * cancelled, so a cancelled caller can never leave a half-installed session on disk.
     */
    suspend fun install(account: AccountBundle) {
        mutex.withLock {
            epoch += 1
            val installed = CurrentAccountSnapshot(
                account = account,
                validatedAtEpochMillis = nowEpochMillis(),
            )
            snapshot = installed
            hasRefreshedAfterColdStart = true
            mutableState.value = account
            persist { accountStorage.store(installed) }
        }
    }

    /**
     * Reconciles the published account with the server when the [trigger] finds it eligible:
     * once after cold start, otherwise only when the last validation is outside [freshness].
     * Concurrent eligible triggers await one shared request.
     */
    suspend fun refresh(trigger: CurrentAccountRefreshTrigger) {
        val pending = mutex.withLock {
            val current = snapshot ?: return
            val eligible = when (trigger) {
                CurrentAccountRefreshTrigger.COLD_START -> !hasRefreshedAfterColdStart
                CurrentAccountRefreshTrigger.FOREGROUND,
                CurrentAccountRefreshTrigger.PROFILE_OPENED,
                -> isStale(current)
            }
            if (!eligible) return
            hasRefreshedAfterColdStart = true
            refreshJob ?: scope.async { performRefresh() }.also { refreshJob = it }
        }
        pending.await()
    }

    private suspend fun performRefresh() {
        try {
            val startEpoch = mutex.withLock { epoch }
            val result = authenticationRepository.getCurrentAccount()
            mutex.withLock {
                result.onSuccess { account ->
                    if (epoch == startEpoch && snapshot != null) {
                        val refreshed = CurrentAccountSnapshot(
                            account = account,
                            validatedAtEpochMillis = nowEpochMillis(),
                        )
                        snapshot = refreshed
                        mutableState.value = account
                        persistQuietly { accountStorage.store(refreshed) }
                    }
                }
                // On failure the snapshot is retained as-is (stale), so a later trigger
                // retries.
            }
        } finally {
            // Even an escaping Throwable must not leave a dead Deferred registered, or every
            // later trigger would await a request that already finished.
            withContext(NonCancellable) { mutex.withLock { refreshJob = null } }
        }
    }

    /**
     * PATCHes the username and, only if no newer operation superseded this one, persists and
     * publishes the submitted username while keeping the current validation time.
     */
    suspend fun updateUsername(username: String): Result<String> {
        val (current, startEpoch) = mutex.withLock { snapshot to epoch }
        if (current == null) {
            return Result.failure(IllegalStateException("No account in storage"))
        }

        return authenticationRepository
            .updateUsername(accountId = current.account.accountId, username = username)
            .map {
                mutex.withLock {
                    val latest = snapshot
                    if (epoch == startEpoch && latest != null) {
                        epoch += 1
                        val updated = latest.copy(account = latest.account.withUsername(username))
                        snapshot = updated
                        mutableState.value = updated.account
                        persistQuietly { accountStorage.store(updated) }
                    }
                }
                username
            }
    }

    /**
     * Clears the published account and its persisted snapshot; results of any still-running
     * refresh or mutation are fenced out and cannot restore the account.
     *
     * A storage failure propagates to the logout use case — silently keeping the snapshot on
     * disk would resurrect the account on the next launch. The delete itself cannot be
     * cancelled, so a caller cancelled mid-logout (e.g. a cleared view model) can no longer
     * leave the snapshot behind.
     */
    suspend fun clear() {
        mutex.withLock {
            epoch += 1
            snapshot = null
            hasRefreshedAfterColdStart = false
            mutableState.value = null
            persist { accountStorage.delete() }
        }
    }

    /** Runs [operation] to completion even if the caller is cancelled; failures propagate. */
    private suspend fun persist(operation: suspend () -> Unit) {
        withContext(NonCancellable) { operation() }
    }

    /**
     * Runs [operation] to completion even if the caller is cancelled, swallowing failures:
     * the server stays authoritative, so the next refresh self-heals a missed write.
     */
    private suspend fun persistQuietly(operation: suspend () -> Unit) {
        withContext(NonCancellable) { runCatching { operation() } }
    }

    private fun isStale(snapshot: CurrentAccountSnapshot): Boolean {
        val validatedAt = snapshot.validatedAtEpochMillis ?: return true
        return nowEpochMillis() - validatedAt >= freshness.inWholeMilliseconds
    }
}
